src/nodes/reducemean.h

```diff
@@ -0,0 +1,302 @@
+/* This file is part of onnx2c.
+ *
+ * TEMPLATE node.
+ * When implementing a new node, use this template
+ * as a starting point.
+ *
+ * This file can be kept as a single .h file with an
+ * in-header implementation, or it can be split into
+ * a .h and a .cc file.
+ *
+ * Replace all occurances of TEMPLATE in this file.
+ * Some representative dummy implementation provided.
+ *
+ * The functions here are callbacks from the onnx2c
+ * framework. See node.h for more documentation.
+ */
+#include "node.h"
+#include "util.h"
+
+#include <algorithm>
+
+namespace toC {
+
+class ReduceMean : public Node {
+	public:
+	ReduceMean() {
+		op_name = "ReduceMean";
+	}
+
+	constexpr static const char* kInputName = "input";
+	constexpr static const char* kOutputName = "reduced";
+
+	/* Attributes */
+	std::vector<int64_t> axes;
+	int keepdims;
+
+	// Mandatory "API" functions towards the rest of onnx2c
+	virtual void parseAttributes( onnx::NodeProto &node ) override;
+	virtual void resolve(void) override;
+	virtual void print(std::ostream &dst) const override;
+	private:
+	void printLocationArray(std::ostream &dst, int indent, int axis, int dims, const char* flatIndexVariable) const;
+};
+
+
+/* Parse attributes, if this node has them. */
+void ReduceMean::parseAttributes( onnx::NodeProto &node )
+{
+	if (onnx_ir_version > 13)
+	{
+		// At version 18, the attributes changed.
+		ERROR("ReduceMean only supported for opset <=13 not " + std::to_string(onnx_ir_version));
+	}
+
+	Tensor* input = get_input_tensor(0);
+
+	// Default to reducing all dimensions
+	for (int i = 0; i < input->data_dim.size(); ++i)
+	{
+		axes.push_back(i);
+	}
+
+	// Default to keeping all dimensions
+	keepdims = 1;
+
+	for( const auto& a : node.attribute() ) {
+		LOG(TRACE) << "Parsing attribute " << a.name() << std::endl;
+		if( a.name() == "axes" )
+		{
+			axes = parse_attribute_ints(a);
+
+			for (int64_t& axis : axes)
+			{
+				while (axis < 0)
+				{
+					axis += get_input_tensor(0)->data_dim.size();
+				}
+			}
+		}
+		else if( a.name() == "keepdims" )
+			keepdims = parse_attribute_int(a);
+		else
+			LOG(ERROR) << "Ignoring attribute " << a.name() << " for node ReduceMean/" << onnx_name << std::endl;
+	}
+}
+
+
+/* Assign input tensors, resolve output tensor shapes, allocate output tensors */
+void ReduceMean::resolve(void)
+{
+	name_input(0, kInputName);
+	Tensor *input = get_input_tensor(0);
+
+	/* Create output tensors.
+	 * Set data dimensions and data type for the created tensors. */
+	Tensor *t = new Tensor;
+	t->data_dim = {};
+
+	if (keepdims)
+	{
+		for (int i = 0; i < input->data_dim.size(); ++i)
+		{
+			if(std::find(axes.begin(), axes.end(), i) != axes.end())
+			{
+				t->data_dim.push_back(1);
+				continue;
+			}
+
+			t->data_dim.push_back(input->data_dim[i]);
+		}
+	}
+	else
+	{
+		for (int i = 0; i < input->data_dim.size(); ++i)
+		{
+			if(std::find(axes.begin(), axes.end(), i) != axes.end())
```

Comment:
I might be confusing myself now, but shouldn't `axes` be a constant tensor here? 
Otherwise, the output size can be run-time dynamic?
If yes, please add an assertion for constness. If no, I'll have a second go at thinking about this :)

Reply:
`axes` is the vector of which axes to reduce which is as an attribute in the opset version I targeted. I think version 18 introduces axes as a tensor. Here I just use it for determining the output shape.

Let me know if that answered your question, I may have misunderstood.

Reply:
An assert similar to that should work. I only use that vector for tracking things during the `onnx2c` run. I didn't add anything dynamic here. It's just an `std::vector` I got from an attribute of the node, not an input tensor.

Reply:
Oh, right. It's an attribute, not an input tensor. So no check needed.